src/i_video.c

```diff
@@ -416,32 +472,65 @@ static inline void I_UpdateRender (void)
     }
     else
     {
-        SDL_RenderCopy(renderer, texture, NULL, NULL);
+        SDL_RenderCopy(renderer, texture, &blit_rect, NULL);
     }
 }
 
+static int DynamicResolution(uint64_t frametime)
+{
+    if (frametime == 0)
+    {
+        return 0;
+    }
+
+    // 1.25 milliseconds for SDL render present
+    double target = (1.0 / refresh_rate) - 0.00125;
+    double actual = frametime / 1000000.0;
+
+    double actualpercent = actual / target;
+
+    #define DRS_MIN_HEIGHT 400
+    #define DRS_DELTA 0.5
+    #define DRS_GREATER (1 + DRS_DELTA)
+    #define DRS_LESS (1 - DRS_DELTA)
+
+    int newheight = 0;
+    int oldheight = video.height;
+
+    if (actualpercent > DRS_GREATER)
+    {
+        double reduction = (actualpercent - DRS_GREATER ) * 0.2;
+        newheight = (int)MAX(DRS_MIN_HEIGHT, oldheight - oldheight * reduction);
+    }
+    else if (actualpercent < DRS_LESS)
+    {
+        double addition = (DRS_LESS - actualpercent) * 0.25;
+        newheight = (int)MIN(renderer_height, oldheight + oldheight * addition);
+    }
```

Comment:
@fabiangreffrath, @mikeday0 I tried to implement your previous suggestions, it works better but still not perfect. Corrections, or other ideas?

Reply:
Much better! Though, sometimes the resolution switches back to 1916x900, whereas my original resolution would be 1920x900. Also, there are some odd-ball resolutions in between: 1840x864 1500x704 1136x536 1672x784. Is there any system behind this?

Reply:
> Though, sometimes the resolution switches back to 1916x900, whereas my original resolution would be 1920x900. 

Currently we only set the height and the width is calculated in `ResetResolution`. I can try to calculate the aspect ratio more precisely.

> Also, there are some odd-ball resolutions in between: 1840x864 1500x704 1136x536 1672x784. Is there any system behind this?

It is trying to optimize the FPS, see the `DynamicResolution` function. The question is whether these resolutions look bad or not, the exact number doesn't matter. Perhaps we should not calculate the height directly, but choose from an array of predefined resolutions.

Reply:
> Currently we only set the height and the width is calculated in `ResetResolution`. I can try to calculate the aspect ratio more precisely.

Hm, if `height = 900` is passed it should always result in the same width, or not?

> Perhaps we should not calculate the height directly, but choose from an array of predefined resolutions.

I think I'd prefer that. It would feel less "alien". 😉

Reply:
> Hm, if `height = 900` is passed it should always result in the same width, or not?

I think rounding up the `height` value to the next multiple of `4` should happen first in that function.

Reply:
> I think I'd prefer that. It would feel less "alien". 😉

Maybe only allow steps of, say, 40px.

Reply:
> Hm, if `height = 900` is passed it should always result in the same width, or not?

There's probably a bug somewhere. I plan to calculate everything using `float`/`double`.

Reply:
This works really well on my system. Nice work, I am super impressed! So far I've been testing with Eviternity Map 26. In the opening area I can get it to switch resolutions by switching between looking at the tower complex and the less-detailed area to the left. The resolution switching is very snappy, and I haven't seen any instances of instability.

Reply:
Thanks, but I suspect you're using pretty powerful hardware and the resolution doesn't drop too low? Currently `DRS_MIN_HEIGHT` is 400p, and there is a  "vibration" noticeable on status bar or weapons rendering when switching resolutions quickly. I'm thinking of setting the minimum to 600p, but then the DRS feature wouldn't be very useful for 1080p displays.

Reply:
i7-4770K, 16 GB of RAM, GTX 770. Good hardware 10 years ago. :)  My display is 1920 x 1200, 60Hz. The lowest resolution I see when running around in Eviternity map 26 is 1152 x 720. (Reported with `-verbose` console output, not the IDRATE cheat.)

Reply:
I guess we need an even more extreme map for testing 😄 I tested it on an old 2011 Core i5 laptop, it drops to a minimum 853x480 (or 400p without aspect ration correction). With a powerful CPU at high resolutions, the "vibration" is much less noticeable.

Reply:
> I guess we need an even more extreme map for testing

Map31 of Antaresian Reliquary might be useful. Maybe Eviternity II has heavier maps, since it is a new release.